# Exercicio Apriori (padroes nos dados)

Por meio do algoritimo apriori, encontre padroes que sejam relevantes na base de dados resultados_rank_nemenyi.csv

Tente explicar um pouco sobre o algoritimo

Bom trabalho!

# Importando Bibliotecas

In [1]:
import pandas as pd
import numpy as np
!pip install apyori
from apyori import apriori

# Resolucao

In [2]:
df = pd.read_csv('census.csv')
df

age          workclass  final-weight    education  education-num  \
0       39          State-gov         77516    Bachelors             13   
1       50   Self-emp-not-inc         83311    Bachelors             13   
2       38            Private        215646      HS-grad              9   
3       53            Private        234721         11th              7   
4       28            Private        338409    Bachelors             13   
...    ...                ...           ...          ...            ...   
32556   27            Private        257302   Assoc-acdm             12   
32557   40            Private        154374      HS-grad              9   
32558   58            Private        151910      HS-grad              9   
32559   22            Private        201490      HS-grad              9   
32560   52       Self-emp-inc        287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loos  hour-per-week  native-country  \
0         Male          2174             0             40   United-States   
1         Male             0             0             13   United-States   
2         Male             0             0             40   United-States   
3         Male             0             0             40   United-States   
4       Female             0             0             40            Cuba   
...        ...           ...           ...            ...             ...   
32556   Female             0             0             38   United-States   
32557     Male             0             0             40   United-States   
32558   Female             0             0             40   United-States   
32559     Male             0             0             20   United-States   
32560   Female         15024             0             40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

In [3]:
# O algoritimo so consegue entender dados no formato str ou obj. 
#Portanto sera necessario criar um DF com os dados a serem utilizados para teste.
def idade(column):
    if column < 18:
        return 'Jovem'
    elif column >= 18 and column <= 25:
        return 'Adulto I'
    elif column >= 26 and column <= 40:
        return 'Adulto II'
    elif column >= 41 and column <= 60:
        return 'Idoso I'
    elif column >= 61 and column <= 90:
        return 'Idoso II'
    else:
        return 'Nao informado'
    
df['age'] = df['age'].apply(lambda x : idade(x))
df.head(7)

age          workclass  final-weight   education  education-num  \
0  Adulto II          State-gov         77516   Bachelors             13   
1    Idoso I   Self-emp-not-inc         83311   Bachelors             13   
2  Adulto II            Private        215646     HS-grad              9   
3    Idoso I            Private        234721        11th              7   
4  Adulto II            Private        338409   Bachelors             13   
5  Adulto II            Private        284582     Masters             14   
6    Idoso I            Private        160187         9th              5   

           marital-status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   

       sex  capital-gain  capital-loos  hour-per-week  native-country  income  
0     Male          2174             0             40   United-States   <=50K  
1     Male             0             0             13   United-States   <=50K  
2     Male             0             0             40   United-States   <=50K  
3     Male             0             0             40   United-States   <=50K  
4   Female             0             0             40            Cuba   <=50K  
5   Female             0             0             40   United-States   <=50K  
6   Female             0             0             16         Jamaica   <=50K

In [4]:
set(df['education-num'])

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}

In [5]:
def educacao(column):
    if column < 10:
        return 'Pouco Estudo'
    else:
        return 'Bem Estudado'
    
df['education-num'] = df['education-num'].apply(lambda x : educacao(x))
df.head(7)

age          workclass  final-weight   education education-num  \
0  Adulto II          State-gov         77516   Bachelors  Bem Estudado   
1    Idoso I   Self-emp-not-inc         83311   Bachelors  Bem Estudado   
2  Adulto II            Private        215646     HS-grad  Pouco Estudo   
3    Idoso I            Private        234721        11th  Pouco Estudo   
4  Adulto II            Private        338409   Bachelors  Bem Estudado   
5  Adulto II            Private        284582     Masters  Bem Estudado   
6    Idoso I            Private        160187         9th  Pouco Estudo   

           marital-status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   

       sex  capital-gain  capital-loos  hour-per-week  native-country  income  
0     Male          2174             0             40   United-States   <=50K  
1     Male             0             0             13   United-States   <=50K  
2     Male             0             0             40   United-States   <=50K  
3     Male             0             0             40   United-States   <=50K  
4   Female             0             0             40            Cuba   <=50K  
5   Female             0             0             40   United-States   <=50K  
6   Female             0             0             16         Jamaica   <=50K

In [6]:
df = df[['age','workclass','education','education-num','marital-status','occupation','relationship','race','sex','native-country','income']]
df.head(7)

age          workclass   education education-num  \
0  Adulto II          State-gov   Bachelors  Bem Estudado   
1    Idoso I   Self-emp-not-inc   Bachelors  Bem Estudado   
2  Adulto II            Private     HS-grad  Pouco Estudo   
3    Idoso I            Private        11th  Pouco Estudo   
4  Adulto II            Private   Bachelors  Bem Estudado   
5  Adulto II            Private     Masters  Bem Estudado   
6    Idoso I            Private         9th  Pouco Estudo   

           marital-status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   

       sex  native-country  income  
0     Male   United-States   <=50K  
1     Male   United-States   <=50K  
2     Male   United-States   <=50K  
3     Male   United-States   <=50K  
4   Female            Cuba   <=50K  
5   Female   United-States   <=50K  
6   Female         Jamaica   <=50K

In [7]:
# Separando uma menor quantidade para ser analisada (Neste caso, onde a maquina nao e' tao boa)
df = df.sample(2500)

In [8]:
transacoes = []
for i in range(df.shape[0]):
    transacoes.append([str(df.values[i,j]) for j in range(df.shape[1])])

In [9]:
transacoes

[['Adulto II',
  ' Private',
  ' Assoc-acdm',
  'Bem Estudado',
  ' Divorced',
  ' Sales',
  ' Unmarried',
  ' White',
  ' Female',
  ' United-States',
  ' <=50K'],
 ['Idoso I',
  ' Private',
  ' HS-grad',
  'Pouco Estudo',
  ' Divorced',
  ' Adm-clerical',
  ' Unmarried',
  ' White',
  ' Female',
  ' United-States',
  ' <=50K'],
 ['Adulto I',
  ' Private',
  ' Bachelors',
  'Bem Estudado',
  ' Never-married',
  ' Sales',
  ' Not-in-family',
  ' White',
  ' Female',
  ' United-States',
  ' <=50K'],
 ['Idoso II',
  ' Private',
  ' Bachelors',
  'Bem Estudado',
  ' Married-civ-spouse',
  ' Sales',
  ' Husband',
  ' White',
  ' Male',
  ' United-States',
  ' >50K'],
 ['Idoso I',
  ' Private',
  ' Bachelors',
  'Bem Estudado',
  ' Never-married',
  ' Tech-support',
  ' Unmarried',
  ' Asian-Pac-Islander',
  ' Female',
  ' India',
  ' <=50K'],
 ['Idoso I',
  ' Private',
  ' HS-grad',
  'Pouco Estudo',
  ' Married-civ-spouse',
  ' Craft-repair',
  ' Husband',
  ' White',
  ' Male',
  ' Unite

Uma vez o Novo DataFrame consolidado e preparado, o que resta e' submete-lo ao algoritmo.

In [10]:
insight = apriori(transacoes,min_support=0.3,min_confidence=0.7)
resultado = list(insight)
print(len(resultado))
resultado

71


[RelationRecord(items=frozenset({' <=50K'}), support=0.7572, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({' <=50K'}), confidence=0.7572, lift=1.0)]),
 RelationRecord(items=frozenset({' United-States'}), support=0.8936, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({' United-States'}), confidence=0.8936, lift=1.0)]),
 RelationRecord(items=frozenset({' White'}), support=0.8528, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({' White'}), confidence=0.8528, lift=1.0)]),
 RelationRecord(items=frozenset({' <=50K', ' Private'}), support=0.5408, ordered_statistics=[OrderedStatistic(items_base=frozenset({' <=50K'}), items_add=frozenset({' Private'}), confidence=0.7142102482831484, lift=1.0338886049263873), OrderedStatistic(items_base=frozenset({' Private'}), items_add=frozenset({' <=50K'}), confidence=0.7828604516502605, lift=1.0338886049263873)]),
 RelationRecord(items=frozenset({' <=50K',

In [11]:
resultado[5]

RelationRecord(items=frozenset({' White', ' <=50K'}), support=0.6324, ordered_statistics=[OrderedStatistic(items_base=frozenset({' <=50K'}), items_add=frozenset({' White'}), confidence=0.8351822503961965, lift=0.9793412879880353), OrderedStatistic(items_base=frozenset({' White'}), items_add=frozenset({' <=50K'}), confidence=0.7415572232645403, lift=0.9793412879880353)])

O Suporte de 0.544 indica que em 54,4% do DataFrame, temos a presenca de quem estudou e escola privada e recebe menos de 50k dolares anuais.

SE recebe menos de 50k 71,80% de chances de ter estudado em escola privada.
SE estudou em escola privada 77,53% de chance de ganhar menos de 50k dolares.

Indicando talvez uma baixa qualidade nas escolas privadas do pais.

In [12]:
resultado[9]

RelationRecord(items=frozenset({' Male', ' Husband'}), support=0.4084, ordered_statistics=[OrderedStatistic(items_base=frozenset({' Husband'}), items_add=frozenset({' Male'}), confidence=1.0, lift=1.497005988023952)])

O Suporte de 0.4084 indica que em 40,84% do DataFrame, temos a presenca de quem estudou menos de 10 anos (Pouco Estudo) recebe menos de 50k dolares anuais.

SE estudou menos de 10 anos, voce tem 88,25% de chance de ganhar menos de 50k dolares.

Observacao clara que quanto menor o estudo mais dificil e' ganhar bem.

In [13]:
resultado[19]

RelationRecord(items=frozenset({' White', ' Private'}), support=0.5936, ordered_statistics=[OrderedStatistic(items_base=frozenset({' Private'}), items_add=frozenset({' White'}), confidence=0.8592935726693689, lift=1.0076144144809673)])

O Suporte de 0.5704 indica que em 57,04% do DataFrame, temos a presenca de Homens e de Brancos no DF.

SE voce e' homem nos EUA, voce tem 86% de chance de ser branco segundo este dataset.

Observacao clara de que a maior parte da populacao americana e' branca.

In [14]:
resultado[28]

RelationRecord(items=frozenset({'Idoso I', ' White'}), support=0.302, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Idoso I'}), items_add=frozenset({' White'}), confidence=0.8511837655016911, lift=0.9981047906914764)])

O ultimo insight a ser extraido sera a relacao nacionalidade americana com o nivel de estudo.
Com dados que aparecem em 48,36% dos dados do dataset (Americano,mais de 10 anos de estudo).

Se observa que quem estuda ao menos 10 anos, tem 90% de chance de ter nasci nos EUA, isso porque a maior parte do dataset sao pessoas que nasceram la, e o restante, a grande maioria sao pessoas que em geral vem de paises pobres como alguns paises latinos ou africanos, o que explicaria a falta de estudo, pois talvez tenha tentado a migracao torcendo por uma vida melhor.